opencog/attention/experimental/ForgettingAgent.h

```diff
@@ -0,0 +1,142 @@
+ * 
+ * It does based on the LTI of Atoms. Low LTI indicates that an atom has not been
+ * of any use for a long time, and additionally, isn't near any other important
+ * atoms. The latter condition is because the ImportanceSpreadingAgent would
+ * otherwise increase the STI of the atom, by moving STI from nearby important
+ * atoms, and increase the likelihood of the atom in question being:
+ *
+ * 1. used in mind processes, and thus
+ * 2. rewarded with stimulus which later gets exchanged for STI. 
+ *
+ * The ForgettingAgent also takes into account the VLTI of an atom. This is a
+ * boolean value that indicates whether the atom in fact is allowed to be
+ * forgotten. This allows a mechanism for ensuring that highly important
+ * information will never be forgotten, even if it's used only very very rarely.
+ *
+ * Forgetting can be tuned via two parameters:
+ *
+ * 1. The ammount of Atoms the AtomSpace should contain.
+ * 2. A range value of what is an accepteable deviation from that. (Allows the agent to run less often and delete more atoms in one go)
+ *
+ * These work in concert to limit how much and what atoms are forgotten.
+ * TODO: Improve Recursive Remove to work with links outher then HebbianLinks
+ */
+class ForgettingAgent : public Agent
+{
+
+private:
+    AtomSpace* a;
+    Logger *log; //!< Logger object for Agent
+
+    /** Set the agent's logger object
+     *
+     * Note, this will be deleted when this agent is.
+     *
+     * @param l The logger to associate with the agent.
+     */
+    void setLogger(Logger* l);
+
+public:
+
+    virtual const ClassInfo& classinfo() const { return info(); }
+    static const ClassInfo& info() {
+        static const ClassInfo _ci("opencog::ForgettingAgent");
+        return _ci;
+    }
+
+    //! Maximum LTI of an atom that can be forgot.
+    AttentionValue::lti_t forgetThreshold;
+
+    //!targetSize of AtomSpace
+    int maxSize;
+    //!acceptable diviation from maxSize;
+    int accDivSize;
+
+    ForgettingAgent(CogServer&);
+    virtual ~ForgettingAgent();
+    virtual void run();
+
+    void forget();
+
+    /** Return the agent's logger object
+     *
+     * @return A logger object.
+     */
+    Logger* getLogger();
+
+}; // class
+
+typedef std::shared_ptr<ForgettingAgent> ForgettingAgentPtr;
+
+/**
+ * Comparison operator for using qsort on a list of Handles.
+ * Returns them with ascending LTI and if equal in LTI,
+ * then sorted ascending by TruthValue.
+ */
+struct ForgettingLTIThenTVAscendingSort
+{
+    ForgettingLTIThenTVAscendingSort(AtomSpace* _a) {};
+
+    bool operator()(const Handle& h1, const Handle& h2)
+    {
+        AttentionValue::lti_t lti1, lti2;
+
+        lti1 = h1->getAttentionValue()->getLTI();
+        lti2 = h2->getAttentionValue()->getLTI();
+        if (lti1 != lti2) return lti1 < lti2;
+        else {
+            float tv1, tv2;
+            tv1 = fabs(h1->getTruthValue()->getMean());
+            tv2 = fabs(h2->getTruthValue()->getMean());
```

Comment:
Yea strange (not my code) but it might have to be changed completely. Cause i don't see a reason why something with a lower Mean should be forgotten earlier. Something with less confidence seems to make more sense. I'll ask in the attention channel on Slack.